bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -2927,8 +2947,17 @@ private static boolean isEqual(Object lhsValue, Object rhsValue, List<ValuePair>
             return false;
         }
 
-        int lhsValTypeTag = getType(lhsValue).getTag();
-        int rhsValTypeTag = getType(rhsValue).getTag();
+        return checkValueEquals(lhsValue, rhsValue, checkedValues, getType(lhsValue), getType(rhsValue));
+    }
+
+    private static boolean checkValueEquals(Object lhsValue, Object rhsValue, List<ValuePair> checkedValues,
```

Comment:
Can we refactor this to only have `Object lhsValue, Object rhsValue, List<ValuePair> checkedValues` as the function parameters, and get the value types inside the method?